### PR TITLE
9961 query tables result parsing

### LINKED ISSUE #9961: `affected_table_names` breaks with users with hyphen in name
When we parse an SQL query to determine affected tables, we use `CDB_QueryTables` which returns an array of tables. If the table needs to be quoted with `"`, we fail to parse it.

E.g:
```
cartodb_user_db=# select CDB_QueryTables('SELECT * FROM "user-name".table');    
                cdb_querytables                                                                                                             
------------------------------------------------                                                                                            
 {"\"user-name\".table"}                                                                                             
(1 row)                                             
```

We fail to remove the outer quotas, and return a wrong result for `affected_tables` from a layer.

cc @oriolbx 

Probably broken since https://github.com/CartoDB/cartodb-postgresql/commit/4be7d4a497e8d1ebabc453a2e94e55e8ce20ea99

### PR DESCRIPTION
**Pull request by javitonino**
Fixes #9961 

Postgres returns an array with the format:
`{table,table}`

There is double quoting going in in here. If the original table name has no special characters (including `"`), it is unquoted in the array. If it has it is quoted once more before going into the array. So `"user-name".table` becomes `"\"user-name\".table"`.

### CODE REVIEW

**Comment by javitonino** [comment]
Please CR @juanignaciosl

File: app/models/carto/layer.rb [hunk 1]
```diff
@@ -48,6 +48,18 @@ def tables_from_table_name_option
       schema_prefix = user_name.present? && table_name.present? && !table_name.include?('.') ? %{"#{user_name}".} : ''
       tables_from_names(["#{schema_prefix}#{table_name}"], user)
     end
+
+    private
+
+    def parse_cdb_querytables_result(result)
+      return [] unless result.present?
+      tables = result[1..-2].split(',').map do |table_name|
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Please document this `[1..-2]`, as it's not obvious at all. Maybe documenting the method expected parameter value would be nice.

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Maybe the method name could be improved, as it doesn't explain the parsing that it does. Maybe `extract_plain_table_name_from_cdb_querytables_result`?

File: spec/models/carto/layer_spec.rb [hunk 2]
```diff
@@ -75,5 +75,39 @@
...
+
+    it 'correctly parses qualified identifiers' do
+      parse('{table,user.pepito}').should eq ['table', 'user.pepito']
+    end
+
+    it 'correctly parses quoted usernames' do
+      parse('{table,"\"user-hyphen\".pepito"}').should eq ['table', '"user-hyphen".pepito']
+    end
+
+    it 'correctly parses quoted usernames and tablenames' do
```
Comment (juanignaciosl) [inline spec/models/carto/layer_spec.rb]: In order to decouple from `cdb_querytables` implementation and make this more robust, I'd also check these values:

Without quoting
`user-hyphen.pepito`
`user-hyphen.table-hyphen`
`pepito.table-hyphen`

Global quoting:
`"pepito.table"`
`"user-hyphen.table-hyphen"`

And the empty quotes:
`""`
`\"\"`


I'm almost sure that your code supports all of them though.

File: app/models/carto/layer.rb (hunk 1 above)
Comment (javitonino) [inline app/models/carto/layer.rb]: Well, taking into account this and the comments about the tests, what about renaming this to `parse_postgresql_array`? In the end, it's what it's actually doing.

About the `[1..-2]` I was thinking of maybe using a regex like `result.gsub(/{(.*)}/, '$1')` (untested) to be more strict and indicate purpose more clearly.

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Well, it's not only a PostgreSQL array but an array with a known meaning/structure. Anyway, I don't want to extend naming discussions, give it a thought and whatever you come up with will be fine.

About the `[1..-2]`, just keep it if you want, I was very misled by wrong idea of mine. You're right thinking about that regexp being better, but up to you.

**Comment by javitonino** [comment]
Ok, after some talks with @juanignaciosl, I did a some changes to use [`unnest`](https://www.postgresql.org/docs/9.5/static/functions-array.html#ARRAY-FUNCTIONS-TABLE) to retrieve the tables in individual rows, instead of trying to parse postgresql arrays. Code is much cleaner now.

Can you review again?

File: app/models/carto/layer.rb [hunk 3]
```diff
@@ -240,12 +240,8 @@ def tables_from_names(table_names, user)
     def affected_table_names(query)
       return [] unless query.present?
 
-      # TODO: This is the same that CartoDB::SqlParser().affected_tables does. Maybe remove that class?
-      query_tables = user.in_database.execute("SELECT CDB_QueryTables(#{user.in_database.quote(query)})").first
-      query_tables['cdb_querytables'].split(',').map { |table_name|
-        t = table_name.gsub!(/[\{\}]/, '')
-        (t.blank? ? nil : t)
-      }.compact.uniq
+      query_tables = user.in_database.execute("SELECT unnest(CDB_QueryTables(#{user.in_database.quote(query)}))")
```
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: Nice :-) No need for `compact`, isn't it?

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: A little out of scope, as it's not new: could this be refactored to a placeholder?

Comment (javitonino) [inline app/models/carto/layer.rb]: AFAIK, there is no way to do this for custom queries on ActiveRecord 😞. You can access private methods or db dependent connection objects to do it. If you know any way to do it, I will gladly replace this mess 💋

Comment (juanignaciosl) [inline app/models/carto/layer.rb]: True 👍

**Pull request merged**